src/esphome/sensor/ppd42x.cpp

```diff
@@ -0,0 +1,122 @@
+
+#include "esphome/sensor/ppd42x.h"
+#include "esphome/log.h"
+
+ESPHOME_NAMESPACE_BEGIN
+
+namespace sensor {
+
+static const char *TAG = "sensor.ppd42x";
+
+void PPD42XComponent::loop() {
+  const uint32_t now = millis();
+  uint32_t duration_pl_02_5 = pulseIn(this->pl_02_5_sensor_->pl_pin_->get_pin(),
+                                      uint8_t(!this->pl_02_5_sensor_->pl_pin_->is_inverted()),
+              this->timeout_us_);
+  uint32_t duration_pl_10_0 = pulseIn(this->pl_10_0_sensor_->pl_pin_->get_pin(),
+                                      uint8_t(!this->pl_10_0_sensor_->pl_pin_->is_inverted()),
+              this->timeout_us_);
+  this->lowpulseoccupancy_02_5_ = this->lowpulseoccupancy_02_5_ + duration_pl_02_5;
+  this->lowpulseoccupancy_10_0_ = this->lowpulseoccupancy_10_0_ + duration_pl_10_0;
+
+  if (now - this->starttime_ > this->timeout_us_) {
+    // last transmission too long ago
+    this->starttime_ = now;
+    parse_data_();
+  }
+}
+float PPD42XComponent::get_setup_priority() const { return setup_priority::HARDWARE_LATE; }
+
+void PPD42XComponent::parse_data_() {
+  switch (this->ctype_) {
+    case PPD42X_TYPE: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
+    }
+    case PPD42X_TYPE_NS: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
+    }
+    case PPD42X_TYPE_NJ: {
+      float pl_02_5_concentration = us_to_pl(this->lowpulseoccupancy_02_5_, this->timeout_us_);
+      float pl_10_0_concentration = us_to_pl(this->lowpulseoccupancy_10_0_, this->timeout_us_);
+      ESP_LOGD(TAG, "Got PM2.5 Concentration %f pcs/L, PM10.0 Concentration: %f pcs/L", pl_02_5_concentration,
+               pl_10_0_concentration);
+      if (this->pl_02_5_sensor_ != nullptr)
+        this->pl_02_5_sensor_->publish_state(pl_02_5_concentration);
+      if (this->pl_10_0_sensor_ != nullptr)
+        this->pl_10_0_sensor_->publish_state(pl_10_0_concentration);
+      break;
```

Comment:
What is the difference between these switch cases? They all look the same to me.

Reply:
we may have only one sensor ? PM2.5 or PM10.0 or both

Reply:
No, I mean why do you differentiate between `this->ctype_` ? All the switch `case`s look the same to me.